src/mapreduce.jl

```diff
@@ -158,6 +158,9 @@ end
     end
 end
 
+@inline _mapreduce(f, op, D::Tuple{<:Any}, init, sz::Size{S}, a::StaticArray) where {S} =
+    _mapreduce(f, op, first(D), init, sz, a)
```

Comment:
Things like `sum(abs2, SA[1 2; 3 4], dims = (1,2))` still won't work, that seems like a bigger project. It would be easy to add a method here to make them give a friendly "not yet supported" error, instead of an "internal" error. But opinions vary as to whether that's a good idea.

Reply:
The change needed is not large though, we can use `Base.mapreducedim!` to generated the code we want:
```julia
@generated function _mapfoldl(f, op, dims::Val{D}, init,
                               ::Size{S}, a::StaticArray) where {S,D}
    _init = init === _InitialValue ? :default : :init
    iter = CartesianIndices(S)
    exprs = fill!(similar(iter, Any, Base.reduced_indices(iter, D)), _init)
    rf(ex, nex) = ex === :default ? :(Base.reduce_first(op, $nex)) : :(op($ex,$nex))
    Base.mapreducedim!(I->:(f(a[$(I.I...)])), rf, exprs, iter)
    return quote
        @_inline_meta
        @inbounds elements = tuple($(exprs...))
        @inbounds return similar_type(a, eltype(elements), Size($(size(exprs))))(elements)
    end
end
```
The current test passed locally, and the influence on compile time seems negligible.